src/main/java/htsjdk/variant/variantcontext/JEXLMap.java

```diff
@@ -11,153 +12,161 @@
+     */
+    public int size() {
+        return jexl.size();
+    }
+
+    public boolean isEmpty() { return this.jexl.isEmpty(); }
+
+    public Boolean put(JexlVCMatchExp jexlVCMatchExp, Boolean aBoolean) {
+        return jexl.put(jexlVCMatchExp, aBoolean);
+    }
+
+    public void putAll(Map<? extends JexlVCMatchExp, ? extends Boolean> map) {
+        jexl.putAll(map);
+    }
+
+    /**
+     * Initializes all keys with null values indicating that they have not yet been evaluated.
+     * The actual value will be computed only when the key is requested via {@link #get(Object)} or {@link #values()}.
+     */
+    private void initialize(Collection<JexlVCMatchExp> jexlCollection) {
+        jexl = new HashMap<>();
+        for (final JexlVCMatchExp exp: jexlCollection) {
+            jexl.put(exp, null);
+        }
+    }
+
+    /**
+     * Evaluates a {@link JexlVCMatchExp}'s expression, given the current context (and setup the context if it's {@code null}).
+     *
+     * @param exp the {@link JexlVCMatchExp} to evaluate
+     *
+     * @throws IllegalArgumentException when {@code exp} is {@code null}, or
+     *                                  when the Jexl expression in {@code exp} fails to evaluate the JexlContext
+     *                                  constructed with the input VC or genotype.
      */
-    private void evaluateExpression(JexlVCMatchExp exp) {
+    private void evaluateExpression(final JexlVCMatchExp exp) {
         // if the context is null, we need to create it to evaluate the JEXL expression
-        if (this.jContext == null) createContext();
+        if (this.jContext == null){
+            createContext();
+        }
+
         try {
             final Boolean value = (Boolean) exp.exp.evaluate(jContext);
             // treat errors as no match
             jexl.put(exp, value == null ? false : value);
-        } catch (Exception e) {
+        } catch (final JexlException.Variable e){
             // if exception happens because variable is undefined (i.e. field in expression is not present), evaluate to FALSE
-            // todo - might be safer if we explicitly checked for an exception type, but Apache's API doesn't seem to have that ability
-            if (e.getMessage() != null && e.getMessage().contains("undefined variable"))
-                jexl.put(exp,false);
-            else
-                throw new IllegalArgumentException(String.format("Invalid JEXL expression detected for %s with message %s", exp.name, (e.getMessage() == null ? "no message" : e.getMessage())));
+            jexl.put(exp,false);
+        } catch (final JexlException e) {
+            // todo - might be better if no exception is caught here but let's user decide how to deal with them; note this will propagate to get() and values()
+            throw new IllegalArgumentException(String.format("Invalid JEXL expression detected for %s with message %s",
```

Comment:
could you replace this throw with 

```
throw new IllegalArgumentexception(String.format("Invalid JEXL expression detected for %s", exp.name), e)
```

 so that the underlying cause is propagated more cleanly?